File: service/app/src/main/java/io/miragon/blueprint/adapter/inbound/operaton/BaseDelegate.java
package io.miragon.blueprint.adapter.inbound.operaton;

import org.operaton.bpm.engine.delegate.DelegateExecution;
import org.operaton.bpm.engine.delegate.JavaDelegate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Base for all JavaDelegates: wraps the work in a try/catch so failures are logged consistently and
 * re-thrown for the engine to handle.
 */
public abstract class BaseDelegate implements JavaDelegate {

    protected final Logger log = LoggerFactory.getLogger(getClass());

    @Override
    public void execute(DelegateExecution execution) throws Exception {
        try {
            executeTask(execution);
        } catch (Exception e) {
            log.error("Error while processing Operaton task '{}'", execution.getCurrentActivityId(), e);
            throw e;
        }
    }

    protected abstract void executeTask(DelegateExecution execution) throws Exception;
}

File: service/app/src/main/kotlin/io/miragon/blueprint/adapter/inbound/operaton/BaseDelegate.kt
package io.miragon.blueprint.adapter.inbound.operaton

import mu.KotlinLogging
import org.operaton.bpm.engine.delegate.DelegateExecution
import org.operaton.bpm.engine.delegate.JavaDelegate

/**
 * Base for all JavaDelegates: wraps the work in a try/catch so failures are logged consistently and
 * re-thrown for the engine to handle.
 */
abstract class BaseDelegate : JavaDelegate {

    protected val log = KotlinLogging.logger {}

    override fun execute(execution: DelegateExecution) {
        try {
            executeTask(execution)
        } catch (e: Exception) {
            log.error(e) { "Error while processing Operaton task '${execution.currentActivityId}'" }
            throw e
        }
    }

    abstract fun executeTask(execution: DelegateExecution)
}
